docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -631,7 +693,9 @@ Builtin variables are available under the `builtin.` prefix. Some examples:
               the template and validating if the result is a valid YAML or JSON value.
     - We will only implement syntax validation. A semantic validation is not possible as the patches will be applied 
       to provider templates. **It’s the responsibility of the ClusterClass author to ensure the patches are semantically valid**.
-
+  - `MachineHealthCheck`
```

Comment:
I think it's intuitive when the "local" ClusterClass.spec.controlPlane.machineHealthCheck overrides the global ClusterClass.spec.machineHealthCheck (similar for MD).

We have kind of a similar pattern where we prioritize template-specific variables over global variables. (it is implemented in that way but it's current not used as the only template-specific variables are builtin variables which are disjunct from the global builtin variables)

P.S. I would expect a local machineHealthCheck to entirely overwrite a global one. I.e. not on a per-field basis, because then you would have to figure out how to unset global MHC fields and this would lead to a weird UX where you might be forced to unset global MHC fields per MD/ControlPlane.


@enxebre @killianmuldoon @ykakarap WDYT?

Reply:
I think we should keep things simple in the first iteration and keep global health check mutually exclusive to health checks at CP/MD level.

We can then relax this constraint at a later stage and introduce merge/override rules in a compatible way (while instead it will be a breaking change to remove it).

Reply:
I've added an issue to cover the idea of an overriding MHC for ClusterClass at https://github.com/kubernetes-sigs/cluster-api/issues/5770